main/lsp/updates.cc

```diff
@@ -126,56 +147,84 @@ void tryApplyDefLocSaver(const core::GlobalState &gs, vector<ast::ParsedFile> &i
         t.tree = ast::TreeMap::apply(ctx, defLocSaver, move(t.tree));
     }
 }
+} // namespace
 
-LSPLoop::TypecheckRun LSPLoop::runSlowPath(LSPFileUpdates updates) const {
+LSPLoop::TypecheckRun LSPLoop::runSlowPath(unique_ptr<core::GlobalState> previousGS, LSPFileUpdates updates,
+                                           bool isCancelable) const {
     ShowOperation slowPathOp(*this, "SlowPath", "Typechecking...");
     Timer timeit(logger, "slow_path");
     ENFORCE(!updates.canTakeFastPath || config.disableFastPath);
     ENFORCE(updates.updatedGS.has_value());
     if (!updates.updatedGS.has_value()) {
         Exception::raise("runSlowPath called with an update that lacks an updated global state.");
     }
-    prodCategoryCounterInc("lsp.updates", "slowpath");
     logger->debug("Taking slow path");
 
     UnorderedSet<int> updatedFiles;
     vector<ast::ParsedFile> indexedCopies;
+    vector<core::FileRef> affectedFiles;
     auto finalGS = move(updates.updatedGS.value());
-    // Index the updated files using finalGS.
-    {
-        core::UnfreezeFileTable fileTableAccess(*finalGS);
-        for (auto &file : updates.updatedFiles) {
-            auto pair = updateFile(move(finalGS), file, config.opts);
-            finalGS = move(pair.first);
-            auto &ast = pair.second;
-            if (ast.tree) {
-                indexedCopies.emplace_back(ast::ParsedFile{ast.tree->deepCopy(), ast.file});
-                updatedFiles.insert(ast.file.id());
+    // Note: Commits can only be canceled if this edit is cancelable, LSP is running across multiple threads, and the
+    // cancelation feature is enabled.
+    const bool committed = finalGS->tryCommitEpoch(
+        updates.versionEnd, isCancelable && multithreadedMode && config.opts.lspCancelableSlowPathEnabled,
```

Comment:
This boolean, passed to `tryCommitEpoch`, controls whether or not the cancelation feature is enabled.